### PR TITLE
Refactor CODE128

### PR DESCRIPTION
**Pull request by SanichKotikov**
Issue https://github.com/lindell/JsBarcode/issues/92
Marked as WIP (work in progress) because I wanna test it in real project before.
@lindell pls, review :)

### CODE REVIEW

**Comment by SanichKotikov** [comment]
Also I'm not sure about deleting of `InvalidStartCharacterException`

**Comment by lindell** [comment]
When looking at it overall it looks excellent :) The constants.js file was very much needed to keep the file more readable and clean.

I will have to take a closer look at all part of the implementation thou. I will see when I have time for that, hopefully tonight, otherwise in the coming days.

**Comment by SanichKotikov** [comment]
I've tested this code as I mentioned before. Works fine for me :)
Also I thinking about implement `quiet zone` for Code128, but as another PR later.

File: src/barcodes/CODE128/auto.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		autoSelectFromAB(string.substring(chars.length), !isA)
+	);
+}
+
+// CODE128C
+function autoSelectFromC(string) {
+	const cMatch = matchSetC(string);
+	const length = cMatch.length;
+
+	if (length === string.length) return string;
```
Comment (lindell) [inline src/barcodes/CODE128/auto.js]: Same as before, make it a curly braces if

File: src/barcodes/CODE128/auto.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+const matchSetALength = (string) => string.match(new RegExp(`^${A_CHARS}*`))[0].length;
+const matchSetBLength = (string) => string.match(new RegExp(`^${B_CHARS}*`))[0].length;
+const matchSetC = (string) => string.match(new RegExp(`^${C_CHARS}*`))[0];
+
+// CODE128A or CODE128B
+function autoSelectFromAB(string, isA){
+	const ranges = isA ? A_CHARS : B_CHARS;
+	const untilC = string.match(new RegExp(`^(${ranges}+?)(([0-9]{2}){2,})([^0-9]|$)`));
+
+	if (untilC) return (
```
Comment (lindell) [inline src/barcodes/CODE128/auto.js]: Same as before, make it a curly braces if

File: src/barcodes/CODE128/constants.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+// Special characters
+export const SHIFT = 98;
+export const START_A = 103;
+export const START_B = 104;
+export const START_C = 105;
+export const MODULO = 103;
+export const STOP = 106;
+
+// Get set by start code
+export const ABC = {
```
Comment (lindell) [inline src/barcodes/CODE128/constants.js]: Maybe this should have a more descriptive name

Comment (SanichKotikov) [inline src/barcodes/CODE128/constants.js]: Maybe `SET_BY_CODE`? Using:
```js
const startSet = SET_BY_CODE[startIndex];
```

**Comment by lindell** [comment]
If you mean what I think you mean with quite zone it should not be needed for every single barcode type. The margin option (left/right) can be used for this.

File: src/barcodes/CODE128/constants.js (hunk 3 above)
Comment (lindell) [inline src/barcodes/CODE128/constants.js]: 👍

**Comment by SanichKotikov** [comment]
I fixed all your comments and removed WIP.

**Comment by lindell** [comment]
Merged into master.

Thanks a lot for the contribution 😃 🎉

**Pull request merged**